### PR TITLE
Fix generation without any shared libraries

### PR DESCRIPTION
**Pull request by vlinkz**
For example with [gsettings-desktop-schemas-rs](https://gitlab.gnome.org/vlinkz/gsettings-desktop-schemas-rs), there is no library bound. At the moment `sys/lib.rs` generates with the following at the end:
```rust
#[link(name = "")]
extern "C" {

}
```
which leads to the error:
```
error[E0454]: link name must not be empty
   --> gsettings-desktop-schemas/sys/src/lib.rs:199:15
    |
199 | #[link(name = "")]
    |               ^^ empty link name
```

### CODE REVIEW

**Comment by sdroege** [comment]
How does it work without a shared library?

**Comment by vlinkz** [comment]
For example with [gsettings-desktop-schemas](https://gitlab.gnome.org/GNOME/gsettings-desktop-schemas), the `.gir` file has the following namespace element
```xml
  <namespace name="GDesktopEnums"
             version="3.0"
             shared-library=""
             c:identifier-prefixes="GDesktop"
             c:symbol-prefixes="g_desktop">
```
since there isn't an actual shared library, only some enums being defined in a header file.

**Comment by sdroege** [comment]
Ah it's a header-only library. That makes sense.

**Comment by sdroege** [comment]
You probably still want a shared library that provides `GType`s for the enums though, but from a Rust point of view that doesn't matter much. Just means you can't pass these enums through `GValue`s (in any language), only as integers.

File: src/codegen/sys/lib_.rs [hunk 1]
```diff
@@ -71,17 +71,19 @@ fn generate_lib(w: &mut dyn Write, env: &Env) -> Result<()> {
...
-    functions::generate_enums_funcs(w, env, &enums)?;
-    functions::generate_bitfields_funcs(w, env, &bitfields)?;
-    functions::generate_unions_funcs(w, env, &unions)?;
-    functions::generate_records_funcs(w, env, &records)?;
-    functions::generate_classes_funcs(w, env, &classes)?;
-    functions::generate_interfaces_funcs(w, env, &interfaces)?;
-    functions::generate_other_funcs(w, env, &ns.functions)?;
-
-    writeln!(w, "\n}}")?;
+    if !env.namespaces.main().shared_libs.is_empty() {
```
Comment (sdroege) [inline src/codegen/sys/lib_.rs:74]: You probably want to add something here that makes things fail if there are any functions but no library name

Comment (vlinkz) [reply src/codegen/sys/lib_.rs:74]: Something like this?

```rust
    if !ns.functions.is_empty() && env.namespaces.main().shared_libs.is_empty() {
        return Err(Error::new(ErrorKind::Other, "No shared library found, but functions found"));
    }
```

Comment (sdroege) [reply src/codegen/sys/lib_.rs:74]: Not sure what the easiest. `ns.functions` only contains the global functions and not the methods IIRC.

Comment (vlinkz) [reply src/codegen/sys/lib_.rs:74]: Maybe something like this:
```rust
    if !ns.package_names.is_empty() && env.namespaces.main().shared_libs.is_empty() {
        return Err(Error::new(ErrorKind::Other, "No shared library found, but package names found"));
    }
```
That way `.gir`s with libraries defined have to have a package name, but headers like this one need to not have one. That way you can't have the combination of having a shared library with functions, but not defining the package.

Comment (sdroege) [reply src/codegen/sys/lib_.rs:74]: Having a package name seems to be orthogonal from have shared libraries. Your header-only library will have a package name (and as such a `pkg-config` file) or not?

I think what just needs checking here is whether there are global functions, or enums/object/flags/records have functions

Comment (vlinkz) [reply src/codegen/sys/lib_.rs:88]: Pushed with it modified to this, which checks all the function lengths. Adding more than just `classes` and `interfaces` might be a bit overkill though.

```rust
    if env.namespaces.main().shared_libs.is_empty()
        && !(records.iter().all(|x| x.functions.is_empty())
            && classes.iter().all(|x| x.functions.is_empty())
            && interfaces.iter().all(|x| x.functions.is_empty())
            && bitfields.iter().all(|x| x.functions.is_empty())
            && enums.iter().all(|x| x.functions.is_empty())
            && unions.iter().all(|x| x.functions.is_empty()))
    {
        return Err(Error::new(
            ErrorKind::Other,
            "No shared library found, but functions were found",
        ));
    }
```

**Review by sdroege — APPROVED** [review]

**Comment by sdroege** [comment]
Thanks!

**Pull request merged**